engine/source_sym_engine.ml

```diff
@@ -61,122 +20,102 @@ let print_result stree =
-         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
-    | Root cst_list ->
-       bprintf buf
-         "Root=\
-          %t%a"
-         (break (ntabs+1))
-         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Node (k_cst_list, cst_opt) ->
+      bprintf buf "Node:{\
+                   %a \
+                   %t} Fallback: %a"
+        (bprint_list ~sep:sep
+           (fun buf (k,cst) -> bprintf buf "%t%a -> %t%a"
+               sep
+               bprint_constructor k
+               (break (ntabs+2))
+               (bprint_tree (ntabs+1)) cst))
+        k_cst_list
+        (break ntabs)
+        (fun buf -> begin match cst_opt with 
+             | Some cst -> fun _ -> bprint_tree (ntabs+1) buf cst
+             | None -> fun _ -> bprintf buf "None" 
+           end) buf
   in
   let buf = Buffer.create 42 in
   bprint_tree 0 buf stree;
   BatIO.write_line BatIO.stdout (Buffer.contents buf)
 
-let sym_exec (source: source_program) =
-  let shares_kst ~wildcard_to_none k ((pattern, expr):clause): clause option =
-    match pattern with
-    | Wildcard when wildcard_to_none = false-> Some (Wildcard, expr)
-    | Constructor (k', plist) when k' = k ->
-      if (List.length plist > 1) then
-        Some (Constructor (Tuple, plist), expr)
-      else
-        Some (plist |> List.hd, expr)
-    | _ -> None
-  in
-  let negate_piop = function
-    | EqInt i -> NeqInt i
-    | NeqInt i -> EqInt i
-    | IsK s -> NotK s
-    | NotK s -> IsK s
-    | IsTrue -> IsFalse
-    | IsFalse -> IsTrue
-    | IsTuple -> NotTuple
-    | NotTuple -> IsTuple
-    | IsNil -> NotNil
-    | NotNil -> IsNil
-    | IsCons -> NotCons
-    | NotCons -> IsCons
-    | EqString s -> NeqString s
-    | NeqString s -> EqString s
-  in
-  let piop_of_constructor = function
-    | Variant s -> IsK s
-    | Int i -> EqInt i
-    | Bool b -> if b then IsTrue else IsFalse
-    | String s -> EqString s
-    | Tuple -> IsTuple
-    | Nil -> IsNil
-    | Cons -> IsCons
+type row = pattern list * source_expr
+
+let sym_exec source =
+  let kst_rows_assoc rows : (constructor * row list) list * row option =
+    let hashtbl = Hashtbl.create 42
+    in
+    let with_wildcard = ref []
+    in
+    let relevant_wildcard_opt = function
+      | [] -> None
+      | wildcards -> Some (wildcards |> List.rev |> List.hd)
+    in
+    let map_to_hashtable = 
+      fun ((pattern, expr):row) ->
+        match List.hd pattern with
+        | Constructor (k, plist) ->
+          let binding = match Hashtbl.find_opt hashtbl k with
+            | Some clause_lst -> (plist, expr)::clause_lst
+            | None -> (plist, expr)::[]
+          in
+          Hashtbl.replace hashtbl k binding
+        | Wildcard -> with_wildcard := (pattern, expr)::!with_wildcard
+        | _ -> ()
+    in
+    List.iter map_to_hashtable rows;
+    let fst = BatHashtbl.bindings hashtbl
+              |> List.map (fun (k, rows) ->
+                  (k, List.rev (!with_wildcard@rows)))
```

Comment:
There is an arity issue here. Let's assume that `N` is the size of rows at the input of the `kst_rows_assoc` function, and `A` is the arity of the constructor `k` in this particular group (hashtable binding). The arity of rows in the `wild_wildcard` submatrix should be `N - 1` (with your current code it is `N`), and the arity of the row in the group is `N - 1 + A`. The two do not match up; to fix this, you should add to each row of `with_wildcard` extra Wildcard patterns; `A` of them, one for each subpattern of `K`.

For exampele, imagine a matrix of the form

```
(K1(p1)      r1)
(K2(p2,q2)   r2)
(_           r3)
```

then you have

```
rows for K1:
  (p1 r1)
rows for K2:
  (p2 q2 r2)
with_wildcard:
  (r3)
```

and you want the final results

```
(p1 r1)
(_  r3)

(p2 q2 r2)
(_  _  r3)
```

Note: turning `_` into `_ _` in the case of `K2` correspond to transforming `_` into the pattern `K2(_, _)` and then applying the same popping+pushing discipline as for the other rows of the group.